[BOLT][binary-analysis] Add initial pac-ret gadget scanner

kbeyls opened the pull request:
This adds an initial pac-ret gadget scanner to the llvm-bolt-binary-analysis-tool.

The scanner is taken from the prototype that was published last year at https://github.com/llvm/llvm-project/compare/main...kbeyls:llvm-project:bolt-gadget-scanner-prototype, and has been discussed in RFC
https://discourse.llvm.org/t/rfc-bolt-based-binary-analysis-tool-to-verify-correctness-of-security-hardening/78148 and in the EuroLLVM 2024 keynote "Does LLVM implement security hardenings correctly? A BOLT-based static analyzer to the rescue?" [Video](https://youtu.be/Sn_Fxa0tdpY) [Slides](https://llvm.org/devmtg/2024-04/slides/Keynote/Beyls_EuroLLVM2024_security_hardening_keynote.pdf)

In the spirit of incremental development, this PR aims to add a minimal implementation that is "fully working" on its own, but has major limitations, as described in the bolt/docs/BinaryAnalysis.md documentation in this proposed commit.  These and other limitations will be fixed in follow-on PRs, mostly based on code already existing in the prototype branch. I hope incrementally upstreaming will make it easier to review the code.

That being said, this patch isn't really small.

I believe that this patch needs 2 "kinds" of reviewers:
1. People familiar with BOLT internals.
2. People familiar with the pac-ret hardening scheme and AArch64.

I expect people familiar with the pac-ret hardening scheme might need to focus on reviewing mostly what is in file
`NonPacProtectedRetAnalysis.cpp`, `AArch64MCPlusBuilder.cpp`  and the unit tests.

I expect people familiar with BOLT might mainly need to focus on the other parts, and the high-level structure of
`NonPacProtectedRetAnalysis.cpp`.

The patch is not very polished currently, but I think it's in the right shape to start getting it reviewed.

Note that I believe that this could also form the basis of a scanner to analyze correct implementation of PAuthABI.

kbeyls committed:
[BOLT][binary-analysis] Add initial pac-ret gadget scanner

This adds an initial pac-ret gadget scanner to the
llvm-bolt-binary-analysis-tool.

The scanner is taken from the prototype that was published last year at
https://github.com/llvm/llvm-project/compare/main...kbeyls:llvm-project:bolt-gadget-scanner-prototype,
and has been discussed in RFC
https://discourse.llvm.org/t/rfc-bolt-based-binary-analysis-tool-to-verify-correctness-of-security-hardening/78148
and in the EuroLLVM 2024 keynote "Does LLVM implement security
hardenings correctly? A BOLT-based static analyzer to the rescue?"
[Video](https://youtu.be/Sn_Fxa0tdpY)
[Slides](https://llvm.org/devmtg/2024-04/slides/Keynote/Beyls_EuroLLVM2024_security_hardening_keynote.pdf)

In the spirit of incremental development, this PR aims to add a minimal
implementation that is "fully working" on its own, but has major
limitations, as described in the bolt/docs/BinaryAnalysis.md
documentation in this proposed commit.  These and other limitations will
be fixed in follow-on PRs, mostly based on code already existing in the
prototype branch. I hope incrementally upstreaming will make it easier
to review the code.

That being said, this patch isn't really small.

I believe that this patch needs 2 "kinds" of reviewers:
1. People familiar with BOLT internals.
2. People familiar with the pac-ret hardening scheme and AArch64.

I expect people familiar with the pac-ret hardening scheme might need to
focus on reviewing mostly what is in file
`NonPacProtectedRetAnalysis.cpp`, `AArch64MCPlusBuilder.cpp`  and the
unit tests.

I expect people familiar with BOLT might mainly need to focus on the
other parts, and the high-level structure of
`NonPacProtectedRetAnalysis.cpp`.

The patch is not very polished currently, but I think it's in the right
shape to start getting it reviewed.

Note that I believe that this could also form the basis of a scanner to
analyze correct implementation of PAuthABI.

kbeyls commented:
@asl FYI: since you mentioned the usefulness of a scanner for checking whether PAuthABI was applied correctly to a binary in your presentation on PAuthABI at the US LLVM dev meeting: The scanner for pac-ret in this PR could be the basis of a scanner that also checks PAuthABI security properties.

kbeyls committed:
format cleanup

asl commented:
Tagging @atrosinenko

smithp35 commented on bolt/docs/BinaryAnalysis.md:
Might be able to simplify this to "Be read-only within the function" or "Be immutable within the function".

kbeyls replied:
I went for "immutable within the function"

smithp35 commented on bolt/docs/BinaryAnalysis.md:
The combined RETAA and RETAB instructions may not obviously write to the return address. Could be worth adding a third category or adding "This includes combined authentication and return instructions."

kbeyls replied:
I went for "this includes combined...."

smithp35 commented on bolt/docs/BinaryAnalysis.md:
The "as containing" in "uses register `x30` as containing the address to return to" is redundant.

Perhaps
"The return instruction `ret` implicitly uses register `x30` as the address to return to."

smithp35 commented on bolt/docs/BinaryAnalysis.md:
Typo "that shows"

smithp35 commented on bolt/docs/BinaryAnalysis.md:
suggest "instructions for a register follows:"

smithp35 commented on bolt/docs/BinaryAnalysis.md:
I think "or a Branch-and-and-link instruction is meant to represent the case where the return address remains in the register so it isn't stored into memory? 

I tried to think of a way of clarifying, but I couldn't come up with anything better. One possible change is to use the Arm ARM term "Branch with Link" so it is probably best to use that rather than "Branch-and-link".

kbeyls replied:
I ended up rephrasing the whole sentence a bit, which didn't include "Branch with Link" anymore. It probably could be improved further....

> For pac-ret, it is roughly that either the return address is
never stored/retrieved to/from memory; or, there are no writes to the register
containing the return address between an instruction authenticating it and a
return instruction using it.

smithp35 reviewed (commented):
A first pass through the comments/documentation. I've not had a chance to go through the code yet.

Although out of scope for this patch, I'm thinking it will be worth working on a mechanism to exclude certain functions from the analysis. For example if someone has used the branch protection attribute to explicitly exclude a function from branch-protection. An obvious starting point would be some kind of exclude list file, possibly using the same format as the sanitizers.

atrosinenko reviewed (commented):
Thank you for the PR! Added a few polishing comments.

atrosinenko commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,520 @@
+        TrackingLastInsts(RegsToTrackInstsFor.size() != 0),
+        Reg2StateIdx(RegsToTrackInstsFor.size() == 0
+                         ? 0
+                         : *std::max_element(RegsToTrackInstsFor.begin(),
+                                             RegsToTrackInstsFor.end()) +
+                               1,
```
```suggestion
                         : *llvm::max_element(RegsToTrackInstsFor) + 1,
```

kbeyls commented on bolt/docs/BinaryAnalysis.md:
Yeah, I don't know what the status is on other platforms. I'll just stick with "Linux distributions".

kbeyls committed:
Address first round of review feedback by Peter Smith

kbeyls commented on bolt/docs/BinaryAnalysis.md:
Thanks! I went for "For each procedure and exception return instruction, the destination register
must have one of the following properties:"

kbeyls commented:
> A first pass through the comments/documentation. I've not had a chance to go through the code yet.

Thank you, I took all your suggestions.

> Although out of scope for this patch, I'm thinking it will be worth working on a mechanism to exclude certain functions from the analysis. For example if someone has used the branch protection attribute to explicitly exclude a function from branch-protection. An obvious starting point would be some kind of exclude list file, possibly using the same format as the sanitizers.

That's a good call! I hadn't thought about that. I'll consider adding that in a later patch.

One of the other later patches planned is to introduce support for passing from the command line which functions should be considered "no-return", to reduce false positive rate. Going with the same format as the sanitizers would probably also be a good idea for that.

kbeyls committed:
Address first round of review feedback by @atrosinenko

kbeyls commented on bolt/include/bolt/Core/MCPlusBuilder.h:
```diff
@@ -550,6 +550,22 @@ class MCPlusBuilder {
+  }
+
+  virtual bool isAuthenticationOfReg(const MCInst &Inst,
+                                     const unsigned RegAuthenticated) const {
```
Thanks for the suggested name change on the function parameter, makes sense.

kbeyls commented on bolt/include/bolt/Passes/NonPacProtectedRetAnalysis.h:
```diff
@@ -0,0 +1,207 @@
+  uint64_t getAddress() const {
+    // 4 bytes per instruction on AArch64;
+    return BB->getFunction()->getAddress() + BB->getOffset() + BBIndex * 4;
+  }
```
I added a FIXME.
I wasn't too bothered about hard-coding "instructions are 4 bytes", since this an AArch64-specific check. Of course, if this would end up getting reused for analyses that can also run for other targets, this needs to be generalized, indeed.

atrosinenko replied:
Looks reasonable, thank you!

Meanwhile, `MCInstReference` looks rather generally-usable while being defined ad-hoc. I guess someday it could be either refactored and promoted to generic BOLT headers or replaced with other generic implementation if/when other implementation arises.

kbeyls commented on bolt/include/bolt/Passes/NonPacProtectedRetAnalysis.h:
```diff
@@ -0,0 +1,207 @@
+raw_ostream &operator<<(raw_ostream &OS, const MCInstInBFReference &);
+
+struct MCInstReference {
+  enum StoredIn { _BinaryFunction, _BinaryBasicBlock };
```
I removed the leading underscores (had to add `class` in a few places where the source code was referring to the classes `BinaryFunction` and `BinaryBasicBlock`).

I went for `ParentKind` instead of `StoredIn`, as the MCInst's parent is either a `BinaryFunction` or a `BinaryBasicBlock`. Does that make sense to you? Or do you think another name is better?

atrosinenko replied:
> I removed the leading underscores (had to add `class` in a few places where the source code was referring to the classes `BinaryFunction` and `BinaryBasicBlock`).

Sorry, missed the fact that underscores were for disambiguation as well. To simplify things a bit more, renaming `BinaryFunction` and `BinaryBasicBlock` enumerator values to something like `FunctionParent` and `BasicBlockParent` should eliminate the need for specifying `class` for non-enumerators.

> I went for `ParentKind` instead of `StoredIn`, as the MCInst's parent is either a `BinaryFunction` or a `BinaryBasicBlock`. Does that make sense to you?

Thank you, `ParentKind` makes sense for me, too.

A bit more nit-picking, feel free to ignore: `ParentKind CurrentLocation` below could be renamed to something like `ParentKind CurrentLocationKind` (or just `LocationKind`...) to clarify that the location is either in `U.BBRef` or `U.BFRef` and `CurrentLocationKind` is the indication which location field is currently active.

kbeyls replied:
> > I removed the leading underscores (had to add `class` in a few places where the source code was referring to the classes `BinaryFunction` and `BinaryBasicBlock`).
> 
> Sorry, missed the fact that underscores were for disambiguation as well. To simplify things a bit more, renaming `BinaryFunction` and `BinaryBasicBlock` enumerator values to something like `FunctionParent` and `BasicBlockParent` should eliminate the need for specifying `class` for non-enumerators.

Thanks, I followed your suggestion here in the latest commit on the PR. 

> > I went for `ParentKind` instead of `StoredIn`, as the MCInst's parent is either a `BinaryFunction` or a `BinaryBasicBlock`. Does that make sense to you?
> 
> Thank you, `ParentKind` makes sense for me, too.
> 
> A bit more nit-picking, feel free to ignore: `ParentKind CurrentLocation` below could be renamed to something like `ParentKind CurrentLocationKind` (or just `LocationKind`...) to clarify that the location is either in `U.BBRef` or `U.BFRef` and `CurrentLocationKind` is the indication which location field is currently active.

I looked a bit at other places in LLVM using "Kind"s to decide which union member is active.
I ended up going for
```
  enum Kind { FunctionParent, BasicBlockParent };
  Kind ParentKind;
```
as that style is in line with many other places in LLVM, and I find it more readable, e.g.
```
    assert(OverwInst.CurrentLocation == MCInstReference::BasicBlockParent);
```
has now become
```
    assert(OverwInst.ParentKind == MCInstReference::BasicBlockParent);
```

kbeyls commented on bolt/include/bolt/Passes/NonPacProtectedRetAnalysis.h:
```diff
@@ -0,0 +1,207 @@
+} // namespace bolt
+} // namespace llvm
+
+#endif
```
I also don't know about any code style rules about newlines at end of files, but I added them here and for `NonPacProtectedRetAnalysis.cpp`.

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,520 @@
+  }
+  virtual ~PacRetAnalysis() {}
+
+  void run() { Parent::run(); }
```
It is redundant indeed, now removed.

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,520 @@
+    EndIndex = BB->size() - 1;
+  const BinaryFunction *BF = BB->getFunction();
+  for (unsigned I = StartIndex; I <= EndIndex; ++I) {
+    uint64_t Address = BB->getOffset() + BF->getAddress() + 4 * I;
```
Agreed, I added a FIXME.

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,520 @@
+          if (BC.MIB->hasAnnotation(Inst, GadgetAnnotationIndex)) {
+            reportFoundGadget(BC, Inst, GadgetAnnotationIndex);
+          }
+        }
```
It does!

atrosinenko commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,520 @@
+  /// track which instructions last wrote to this register.
+  std::vector<uint16_t> Reg2StateIdx;
+
+  bool trackReg(MCPhysReg Reg) {
```
> But now that I'm writing this comment, I'm wondering if this could be misinterpreted as meaning "start tracking this register" rather than "should I be tracking this register"...

Agree, `doTrackReg` sounds more like an action to me.

> Probably `isTrackingReg` is better indeed. Will change that in my next update.

Agree, this sounds unambiguous.

kbeyls committed:
doTrackReg -> isTrackingReg

jacobbramley commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -148,6 +149,68 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+
+    case AArch64::AUTIA:
+    case AArch64::AUTIB:
+    case AArch64::AUTDA:
+    case AArch64::AUTDB:
```
Should we include data authentication for this analysis? To branch (return or otherwise) to a data-authenticated pointer is surely a programming error, but perhaps a warning would be useful.

kbeyls replied:
Hmmm. Good question...
The analysis as currently implemented checks "did any authentication happen on the value in this register that is going to be used in a return instruction", rather than "did an _instruction_ authentication happen".

If no strong indication emerges why this could hide a gap in pac-ret security hardening implementation that will be left undetected through "normal" correctness testing (i.e. checking that the binary processes inputs and translates it to outputs as expected), I think it may be best to leave this idea as a potential low-priority improvement to add later.

jacobbramley commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -148,6 +149,68 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
     return false;
   }
 
+  MCPhysReg getAuthenticatedReg(const MCInst &Inst) const override {
```
This seems to be missing `AUTI[AB]SPPC`, `AUTI[AB]SPPCR` and corresponding `RETA[AB]*` variants.

kbeyls replied:
Thanks for noticing! The original code here was written before these instructions were fully defined...
I've added them now in commit `f44f9bf`, but still need to write regression tests for these instructions.

kbeyls replied:
The regression tests have been added in the latest commit, I believe the work related to this comment is now complete.

jacobbramley commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -148,6 +149,68 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+      return AArch64::X17;
+    case AArch64::ERETAA:
+    case AArch64::ERETAB:
+      return AArch64::LR;
```
I'm not very aware of all the details, but don't `ERET` instructions return to `ELR`? Tracking writes to that might actually be simpler — presumably an `MSR` — but would need a different mechanism.

kbeyls replied:
I think you're right, thanks for pointing this out.
I investigated this a bit, and came to the following preliminary conclusions.

`ERETAA`/`ERETAB`/`ERET` instructions return to the address stored in one of
the system registers `ELR_EL1`, `ELR_EL2` or `ELR_EL3`, depending on the
current "Exception Level" state.

I'm not aware of any examples of use of `ERETA{A|B}` in open source
software currently. My understanding is that it could be used as follows to create a
pac-ret-like hardening for `ERET`:

```
       exception_entry:
         MRS x0, elr_el1 // Could be elr_el2, elr_el3 depending on whether
                         // this code is designed to run at EL1, EL2 or EL3.
         PACIA x0, SP
         STR   x0, [some place]
         ...
         LDR   x0, [some place]
         MSR   elr_el1, x0
         ERETAA
```

Assuming my understanding above is correct, this makes it impossible to
write a fully static "pac-ret"-like checker for exception returns,
because the static analyzer doesn't know whether the code will run at
EL1, EL2 or EL3.  Therefore, the static analyzer can't know whether to
check for writes to `elr_el1`, `elr_el2` or `elr_el3`.

Also, LLVM does not model the system registers `ELR_EL1`, `ELR_EL2` and `ELR_EL3` as regular registers. Presumably, tracking changes to system registers would require quite substantial changes to the current implementation of the analyzer.

I decided to handle ERETs in the analyzer by producing a warning when encountered, which states that the analyzer cannot analyze ERETs.

This was implemented in the latest commit f44f9bf on this PR.

This change results in the non-pac-ret-protected-return analyzer now
needing to produce 2 different kinds of diagnostics.

I found that this is nicely modelled by creating a class hierarchy
for these diagnostics.

However, it seems that MCAnnotations (which was the way diagnostics to be produced were stored during analysis) currently do not work (well or at
all) when trying to store any object from a class hierarchy, rather than
an object with fixed type.
I realized that it was a bit strange that the diagnostics
that need to be produced are stored using MCAnnotations at all...
So, the latest commit 
f44f9bf redesigns that aspect so that the diagnostics that need to be
produced are not stored as MCAnnotations on an MCInst at all...

Overall, it seems like this is a cleaner design and a better example for
any future binary analyses.

jacobbramley commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -148,6 +149,68 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+      return getNoRegister();
+    case AArch64::RETAA:
+    case AArch64::RETAB:
+    case AArch64::ERETAA:
+    case AArch64::ERETAB:
+    case AArch64::ERET:
```
Again, `ELR`?

jacobbramley commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -148,6 +149,68 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+    switch (Inst.getOpcode()) {
+    case AArch64::RET:
+      // There should be one register that the return reads, and
+      // that's the one being used as the jump target?
```
? -> .

jacobbramley commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -148,6 +149,68 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+    case AArch64::RET:
+      // There should be one register that the return reads, and
+      // that's the one being used as the jump target?
+      for (unsigned OpIdx = 0, EndIdx = Inst.getNumOperands(); OpIdx < EndIdx;
+           ++OpIdx) {
+        const MCOperand &MO = Inst.getOperand(OpIdx);
+        if (MO.isReg())
+          return MO.getReg();
+      }
+      return getNoRegister();
```
This seems very defensive. Is it not an error for a `RET` not to have exactly one register operand?

kbeyls replied:
Yeah, this is a left-over from when I initially wrote the prototype and rather than spending the effort of looking up which operand is the register on a `RET` instruction as encoded in MC, I wrote this loop. I replaced this loop and the comment before now with `return Inst.getOperand(0).getReg();`

jacobbramley reviewed (commented):
Thanks @kbeyls!

This is just a first pass of the implementation, mostly just nits. I'll have a look through the unit tests next week, but I'm already fairly confident after having used it a few times on real code.

kbeyls committed:
ParentKind->Kind; CurrentLocation->ParentKind

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,517 @@
+// exactly the registers containing values that should not be trusted (as they
+// could have changed since the last time they were authenticated). For pac-ret,
+// any return instruction using such a register is a gadget to be reported. For
+// PAuthABI, any indirect control flow using such a register should be reported?
```
I'm not fully sure, but I think in PAuthABI, there might be pointers that are signed and authenticated that are not "indirect control flow pointers", for example vtable pointers. Since I'm not sure about this (haven't investigated in detail), I prefer to leave the question mark for now, to indicate some level of "this is probably not the full story".

kbeyls replied:
Yeah, makes sense. I've ended up just improving the comment so that it didn't need a question mark at the end, and also doesn't need a "FIXME" marker...

kbeyls committed:
Address the easy-to-address feedback by Jacob

atrosinenko commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,517 @@
+    if (TrackingLastInsts) {
+      for (auto WrittenReg : Written.set_bits()) {
+        if (isTrackingReg(WrittenReg)) {
+          Next.LastInstWritingReg[reg2StateIdx(WrittenReg)] = {};
+          Next.LastInstWritingReg[reg2StateIdx(WrittenReg)].insert(&Point);
```
[nit] Strictly speaking, this seems to be equivalent to
```cpp
  Next.LastInstWritingReg[reg2StateIdx(WrittenReg)] = {&Point};
```
though, I guess it would be even less readable. On the other hand, replacing ` = {}` with `.clear()` looks like an improvement for readability:
```suggestion
          Next.LastInstWritingReg[reg2StateIdx(WrittenReg)].clear();
          Next.LastInstWritingReg[reg2StateIdx(WrittenReg)].insert(&Point);
```

kbeyls replied:
I thought simply writing
```
Next.LastInstWritingReg[reg2StateIdx(WrittenReg)] = {&Point};
```
in a single statement is clearer than doing it in 2 steps. I've also slightly rewritten the loop nest further so that it is now quite a bit easier to read. See 0480ce7

atrosinenko reviewed (commented):
Thanks @kbeyls! Here are a few more comments, most of them are trivial nit-picking.

atrosinenko commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,681 @@
+// RUN: %clang %cflags -march=armv8.3-a -mbranch-protection=pac-ret %s %p/../../Inputs/asm_main.c -o %t.exe
+// RUN: llvm-bolt-binary-analysis --scanners=pacret %t.exe 2>&1 | FileCheck -check-prefix=CHECK --allow-empty %s
```
Redundant options:
```suggestion
// RUN: llvm-bolt-binary-analysis --scanners=pacret %t.exe 2>&1 | FileCheck %s
```

atrosinenko commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,681 @@
+        bl      g
+        add     x0, x0, #3
+        ldp     x29, x30, [sp], #16
+        autiaz
```
```suggestion
        autibz
```

atrosinenko commented on bolt/test/binary-analysis/AArch64/gs-pacret-multi-bb.s:
```diff
@@ -0,0 +1,46 @@
+// CHECK-NEXT:    The return instruction is     {{[0-9a-f]+}}:       ret
+// CHECK-NEXT:    The 1 instructions that write to the return register after any authentication are:
+// CHECK-NEXT:    1.     {{[0-9a-f]+}}:      ldp     x29, x30, [sp], #0x10
+
```
A test case similar to shrink-wrapping could probably be interesting, too:
```
        cbz     x0, 1f
        hint    #25
        stp     x29, x30, [sp, #-16]!
        ldp     x29, x30, [sp], #16
        autiasp
1:
        ret
```

kbeyls replied:
Thanks, I added this now in commit a05ae62

atrosinenko commented on bolt/test/binary-analysis/AArch64/gs-pacret-multi-bb.s:
```diff
@@ -0,0 +1,46 @@
+// RUN: %clang %cflags -march=armv8.3-a -mbranch-protection=pac-ret %s %p/../../Inputs/asm_main.c -o %t.exe
+// RUN: llvm-bolt-binary-analysis --scanners=pacret %t.exe 2>&1 | FileCheck -check-prefix=CHECK --allow-empty %s
```
Redundant options:
```suggestion
// RUN: llvm-bolt-binary-analysis --scanners=pacret %t.exe 2>&1 | FileCheck %s
```

kbeyls committed:
Handle erets by producing diagnostic that analysis cannot analyze them correctly.

ERETAA/ERETAB/ERET instructions return to the address stored in one of
the system registers ELR_EL1, ELR_EL2 or ELR_EL3, depending on the
current "Exception Level" state.

I'm not aware of any examples of use of ERETA{A|B} in open source
software currently. My understanding is that could be used as follows to create a
pac-ret-like hardening for ERET:

       exception_entry:
         MRS x0, elr_el1 // Could be elr_el2, elr_el3 depending on whether
                         // this code is designed to run at EL1, EL2 or EL3.
         PACIA x0, SP
         STR   x0, [some place]
         ...
         LDR   x0, [some place]
         MSR   elr_el1, x0
         ERETAA

Assuming my understanding above is correct, this makes it impossible to
write a fully static "pac-ret"-like checker for exception returns,
because the static analyzer doesn't know whether the code will run at
EL1, EL2 or EL3.  Therefore, the static analyzer can't know whether to
check for writes to elr_el1, elr_el2 or elr_el3.

Therefore, in this commit, changes are made to produce a warning when
any of the ERET* instructions are encountered, stating that the analyzer
can't analyze them.

This in turn results in the non-pac-ret-protected-return analyzer now
needing to produce 2 different kinds of diagnostics.

I found that this was most nicely modelled by creating a class hierarchy
for these diagnostics.

However, it seems that MCAnnotations currently do not work (well or at
all) when trying to store any object from a class hierarchy, rather than
an object with fixed type.
At which point I realized that it was a bit strange that the diagnostics
that need to be produced are stored using MCAnnotations at all...
So, this patch redesigns that so that the diagnostics that need to be
produced are not stored as MCAnnotations on an MCInst at all...

Overall, it seems like this is a cleaner design and a better example for
any future binary analyses.

kbeyls commented on bolt/include/bolt/Passes/NonPacProtectedRetAnalysis.h:
```diff
@@ -0,0 +1,266 @@
+  }
+};
+
+struct NonPacProtectedRetAnnotation {
```
With the changes in commit f44f9bf, the code now has quite a few different classes (and variables?) etc with long names containing the string "NonPacProtectedRet".
Could this be improved by introduced a namespace `NonPacProtectedRet` or `NonPacProtectedRetAnalysis`, so that we can remove that part from class names for classes in the namespace. This should make the code substantially easier to read?

kbeyls replied:
Now done in commit  8d4ac8e

kbeyls committed:
Improve comment related to pauthabi checking

kbeyls committed:
Make a few trivial-to-apply improvements suggested by @atrosinenko

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,517 @@
+  /// must compute which the last set of instructions writing to it are.
+  const std::vector<MCPhysReg> RegsToTrackInstsFor;
+  const bool TrackingLastInsts;
+  /// Reg2StateIdx maps Register to the index in the vector used in State to
+  /// track which instructions last wrote to this register.
+  std::vector<uint16_t> Reg2StateIdx;
```
If we didn't use this indirection,
```
std::vector<SmallPtrSet<const MCInst *, 4>> LastInstWritingReg;
```
in the `State` struct, which is stored as an annotation on every instruction `MCInst` in the dataflow analysis, would have to become a map from a register to the set of instructions that last wrote to that register:
```
std::map<MCPhysReg, SmallPtrSet<const MCInst *, 4>> LastInstWritingReg;
```

Does that help to explain why I went for an extra indirection? (to avoid std::map which I assume is way more costly than the indirection + vector approach).

Or am I misunderstanding your comment here?

That being said, it looks like maybe I could use [`IndexedMap`](https://llvm.org/docs/ProgrammersManual.html#llvm-adt-indexedmap-h) to use
```
IndexedMap<SmallPtrSet<const MCInst *, 4>, MCPhysReg2IndexFunctor> LastInstWritingReg;
```

I'll experiment and explore a bit to see if that would make the code clearer.

IIUC, [this review feedback](https://github.com/llvm/llvm-project/pull/122304#discussion_r1913001920) from @atrosinenko, it suggests an alternative approach to make the code more readable, which may well be the easiest way to make the code more readable.

jacobbramley replied:
My assumption was that `SmallPtrSet` is small enough that a ~32-entry `std::vector<SmallPtrSet<...>>`, indexed by `MCPhysReg` rather than `Reg2StateIdx[MCPhysReg]`, would have an acceptable cost, and may be faster in practice. However, either of `IndexedMap` or `lastWritingInsts` would be good (or better) anyway.

kbeyls replied:
LLVM actually models *895* different MCPhysReg registers for AArch64. So, it would be a 895-entry `std::vector`...
895 seems like an excessively large number, but each register (including Neon, SVE, NZCV, ...) gets modelled, and every way an instruction may write to only part of a register gets modelled as a different `MCPhysReg`... 
If you're interested to see the whole list: it's in a tablegen-generated file called `AArch64GenRegisterInfo.inc`. I found an online copy at https://github.com/gdbinit/MachOView/blob/master/capstone/arch/AArch64/AArch64GenRegisterInfo.inc

kbeyls replied:
It seems that I can't use `IndexedMap` here because it requires the indexing function to be compile-time defined. We need the indexing to be defined at run-time, depending on which registers need to be tracked.
I did implement the suggestion from @atrosinenko, and think that together with a few other minor cleanups in commit 0480ce7 , it makes the code quite a lot easier to read.

kbeyls committed:
Add regression tests for the instructions added in pauth-lr: AUTI{A,B}SPPC{i,r}, RETI{A,B}SPPC{i,r}, AUTI{A,B}171615.

jacobbramley commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,676 @@
+        .globl  f1
+        .type   f1,@function
+f1:
+        hint    #25
```
We use `autiasp` elsewhere, so can we write `paciasp` here?

kbeyls commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -2022,6 +2022,8 @@ let Predicates = [HasPAuthLR] in {
     def AUTIASPPCr : SignAuthOneReg<0b00001, 0b100100, "autiasppcr">;
     def AUTIBSPPCr : SignAuthOneReg<0b00001, 0b100101, "autibsppcr">;
+  }
+  let Defs = [X17], Uses = [X15, X16, X17] in {
```
This bug fix in the description of which registers instructions {PAC|AUT}I{A|B}171615 read and write should be upstreamed in a separate PR.

kbeyls replied:
The separate PR for this fix is now created: #123354

kbeyls replied:
... and merged.

jacobbramley commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,851 @@
+// CHECK-NEXT: {{[0-9a-f]+}}:   mov     x16, x30
+// CHECK-NEXT: {{[0-9a-f]+}}:   autiasp
+// CHECK-NEXT: {{[0-9a-f]+}}:   ret     x16
+        ret     x16
```
It might be useful to have a positive test returning to a non-`lr` register.

kbeyls replied:
Now added in [35dfd8d](https://github.com/llvm/llvm-project/pull/122304/commits/35dfd8def3ff4fe86c7a4bc585bdddfb9a2dcde7).

It was useful indeed, as it helped to surface a (so far undocumented/unknown) class of false negative:
When the code moves a properly authenticated value from one register to another register, and then uses that other register to return to.

I haven't encountered an example of this in any of the many binaries I scanned across linux distros.
But maybe this case might be more frequently encountered for pauthabi @atrosinenko ?

I'm not sure currently how the scanner would have to be adapted to handle this case well.
Maybe instead of tracking which registers have been written to, it should track which registers are known to be "properly authenticated"? The rules for "properly authenticated" might be:
* At the start of a function, all registers would be assumed to be properly authenticated.
* Writing to a register with a value loaded from memory invalidates the "properly authenticated" property.
* Any "pure" data-processing instructions would have their destination registers "properly authenticated" iff all of the input registers are "properly authenticated"? (Verifying whether this is correct for all possible instructions would be a lot of work, so maybe it's not worthwhile to implement this unless we see examples of these code sequences from real binaries?)

jacobbramley replied:
Ah, I thought it was a known limitation, sorry!

It's probably infeasible to address it in the general case. Maybe `mov`s can be handled, but then if it's just a `mov`, why wasn't it authenticated directly into the target register? There might be some hint-space reason for it (`autia1617; mov x0, x16; <use x16 for something else>; blr x0`).

For other data-processing operations, it's never really simple:

```
# Valid instruction addresses never have bits <1:0> set, so we can mask them
# freely, or store metadata that's expected to match for authentication. With
# I-side TBI enabled, it should be possible to do the same with the top byte,
# but BOLT probably can't know about that.
autia x10, x10
and x10, x10, #0xffff'ffff'ffff'fffc
ret x10
```

I'd guess that real binaries just won't do anything like this, and if they do, they could be added as exceptions when they come up. PAuthABI is worth thinking about, though.

kbeyls replied:
Right. I think that for this initial commit, we should not try and come up with heuristics to (partially) solve this, and instead leave any potential improvements to future work, when it is demonstrated those heuristics are needed in practice.

jacobbramley commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,851 @@
+        ret
+        .size f_autib171615, .-f_autib171615
+
+// TODO: add test to see if registers clobbered by a call are picked up.
```
That's a good point. Potentially, any branch to an unknown (i.e. non-local) function should be assumed to clobber all the registers (caller-saved or not). I don't recall seeing that in the implementation.

kbeyls replied:
Indeed, this was not taken into account.
Now fixed in c1103cf , with associated test cases.

jacobbramley commented on bolt/test/binary-analysis/AArch64/gs-pacret-multi-bb.s:
```diff
@@ -0,0 +1,46 @@
+// RUN: llvm-bolt-binary-analysis --scanners=pacret %t.exe 2>&1 | FileCheck %s
+
+
+// Verify that we can also detect gadgets across basic blocks
```
I think the implementation attempts to correctly handle cases like this:

```
    cbnz x0, 1f
    autibsp
    b 2f
1:  autiasp
2:  ret
```

That is, two (or more) basic blocks that all authenticate.

kbeyls replied:
Thanks, this is now added in commit a05ae62

jacobbramley reviewed (commented):
Comments on the unit tests.

Thanks!

kbeyls committed:
Introduce lastWritingInsts helper function to improve readability

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,517 @@
+  uint16_t reg2StateIdx(MCPhysReg Reg) const {
+    assert(Reg < Reg2StateIdx.size());
+    return Reg2StateIdx[Reg];
+  }
```
Thanks, that makes a lot of sense. I've introduced such a helper function in 
0480ce7

kbeyls committed:
Add a few more regression tests

kbeyls committed:
Add a test for correct uses of non-LR return. This also adds a FIXME for a false positive category

kbeyls committed:
Assume a call may clobber all registers.

kbeyls committed:
Make many class names much more readable by introducing namespace to avoid repeating NonPacProtectedRet all over the place

jacobbramley commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,914 @@
+// CHECK-NEXT:  This happens in the following basic block:
+// CHECK-NEXT: {{[0-9a-f]+}}:   bl      g@PLT
+// CHECK-NEXT: {{[0-9a-f]+}}:   ret x19
+        // x19, according to the Arm ABI (AAPCS) is a callee-saved register.
+        // Therefore, if function g respects the AAPCS, it should not write
+        // anything to x19. However, we can't know whether function g actually
+        // does respect the AAPCS rules, so the scanner should assume x19 can
+        // get overwritten, and report a gadget if the code does not properly
+        // deal with that.
```
Not only that, but also if it's callee-saved then there's a good chance that it's been saved on a stack at some point, so this is equivalent to having been loaded from memory even the callee does attempt to restore it.

kbeyls replied:
Ah, yes, excellent point, thanks. I've improved the comment both here in the test case and in the analysis implementation to try and cover that.

kbeyls committed:
Improve comment + fix formatting issue

kbeyls commented:
I think I've now addressed all feedback I received.
Do people think this is now OK to merge?

smithp35 commented:
I'm happy my comments have been addressed. Another reviewer of kind "People familiar with the pac-ret hardening scheme and AArch64."

atrosinenko reviewed (commented):
Sorry for the delay, here are mostly trivial comments.

kbeyls committed:
Many small improvements

kbeyls commented on bolt/docs/BinaryAnalysis.md:
I went for "appropriate" rather than "different". Hopefully that eliminates the potential confusion with using "different"?

atrosinenko replied:
This sounds good to me, thank you!

kbeyls commented on bolt/docs/BinaryAnalysis.md:
Yeah, the only option that seemed reasonable in "standard" markdown is a block quote, so I went for that.

atrosinenko replied:
This seems to be [rendered](https://github.com/llvm/llvm-project/blob/6f8b6158ec107daa48da0dab3430b1ff4cde54c4/bolt/docs/BinaryAnalysis.md#pac-ret-analysis) as intended (by Github, at least), thank you!

kbeyls commented on bolt/docs/BinaryAnalysis.md:
Should now be fixed in the latest commit

kbeyls commented on bolt/docs/BinaryAnalysis.md:
Hopefully now corrected to having `ret x16`.

kbeyls commented on bolt/include/bolt/Passes/NonPacProtectedRetAnalysis.h:
```diff
@@ -0,0 +1,266 @@
+
+struct MCInstInBFReference {
+  BinaryFunction *BF;
+  uint32_t Offset;
```
Should now be fixed in the latest commit

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,543 @@
+  /// will probably have length 1 in the second run.
+  std::vector<SmallPtrSet<const MCInst *, 4>> LastInstWritingReg;
+  State() {}
+  State(uint16_t NumRegs, uint16_t NumRegsToTrack)
```
The value for NumRegs comes from [MCRegisterInfo::getNumRegs](https://llvm.org/doxygen/MCRegisterInfo_8h_source.html#l00414), which returns an `unsigned` type.
So yes, it fully makes sense to use unsigned here...

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,543 @@
+public:
+  std::vector<MCInstReference>
+  getLastClobberingInsts(const MCInst Ret, BinaryFunction &BF,
+                         const BitVector &DirtyRawRegs) const {
```
Would `NonAutClobRegs` be a better name. That name would make it consistent with the name for it used in the `State` object?

atrosinenko replied:
I think, the problem with `NonAutClobRegs` name here is that while this variable is derived from `State.NonAutClobRegs`, they are not equal.

Maybe something like `UsedDirtyRegs` reflects the intention better? Here `DirtyRegs` roughly summarizes `NonAutClobRegs` and this set of regs is then intersected with those registers that are `Used` by the return instruction. Of course, when printed by `computeDfState` for the first time, it is indeed better to label it `NonAutClobRegs`:
```cpp
        BitVector UsedDirtyRegs = PRA.getStateAt(Inst)->NonAutClobRegs;
        LLVM_DEBUG({
          dbgs() << "  NonAutClobRegs at Ret: ";
          RegStatePrinter RSP(BC);
          RSP.print(dbgs(), UsedDirtyRegs);
          dbgs() << "\n";
        });
        UsedDirtyRegs &= BC.MIB->getAliases(RetReg, /*OnlySmaller=*/true);
        // ...
```

kbeyls replied:
That all makes sense to me, I followed your suggestions in the latest commit.

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,543 @@
+        RegsToTrack.push_back(R);
+      PacRetAnalysis PRWIA(BF, AllocatorId, RegsToTrack);
+      FAR = computeDfState(PRWIA, BF, AllocatorId);
+    }
+    {
```
Makes sense, I added a comment stating why we need a lock.

atrosinenko reviewed (commented):
Thank you @kbeyls ! I added a few more comments covering all the files except `NonPacProtectedRetAnalysis.cpp` - in other respects these files look good to me in their current state.

kbeyls committed:
More small improvements

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,543 @@
+
+void Annotation::print(raw_ostream &OS) const { OS << *this; }
+void Gadget::print(raw_ostream &OS) const { OS << *this; }
+void GenDiag::print(raw_ostream &OS) const { OS << *this; }
```
Yes, it looks like you're right.
When I looked into testing this, I found out that these `print` and related `operator<<` functions are not called anywhere, so I decided to "fix" this by simply removing these functions.

kbeyls commented on bolt/docs/BinaryAnalysis.md:
thanks for catching this!

kbeyls commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,918 @@
+// CHECK-NOT: function f_nonx30_ret_ok
+        ret     x16
+        .size f_nonx30_ret_ok, .-f_nonx30_ret_ok
+
```
I haven't really thought this through in detail, but my current feel is that there really isn't anything special about `LR` (or `x30`) versus other registers at function entry.

Sure, in many cases, the value in `LR` will have been produced by executing the `BL` instruction that jumps to this function. But, for example, when the function is called through a tail call, the value of the LR register might be a "non-authenticated value that comes from memory", for example:

```
  str x30, [sp]
  ldr x30, [sp]
  b   f
  ...

f:
  ret x30
```

This form isn't really different from the following form (which probably wouldn't show up in real code, unless in a code generator following very different ABI rules...)
```
  adr x1, continuation_address
  b   f
continuation_address:
  nop
  ...

f:
  ret x1
```

I'm now starting to wonder if the first example above is an example of a false negative with the current implementation of the pac-ret scanner in the PR...

Thinking about how this false negative could be fixed, my immediate ideas are currently limited to:
* Assume on every call (tail call or otherwise) that the value in register `x30` will likely be used to return, and therefore must not be a non-authenticated value coming from memory. This could lead to false positives when the function that is called never uses `x30` to return...
* Alternatively, do 2 separate analyses:
   * first do an analysis for each function, recording which of the register values at function start may get used as a code (or also data?) pointer without first being authenticated.
   * In a second analysis, extend the current pac-ret analysis to also check at calls to a particular function, that there is no intersection between the set of `NonAutClobRegs` and the set of registers computed for that function in the first analysis.
  Of course, this would require knowing at every call, also indirect calls, which the set of possible functions called would be, which seems unlikely to be true in general.

atrosinenko replied:
> Assume on every call (tail call or otherwise) that the value in register `x30` will likely be used to return, and therefore must not be a non-authenticated value coming from memory. This could lead to false positives when the function that is called never uses `x30` to return...

Something like this is already implemented in LLVM (though, I'm not sure it is explicitly requested by any ABI-related specification): before emitting a `TCRETURN*` pseudo instruction after LR was authenticated, one of a number of checker sequences can optionally be emitted to make sure that LR was successfully authenticated. The main kinds of checks currently implemented are:
* just perform a load using the pointer in question (can be incompatible with execute-only memory ranges)
* compare bits 61 and 62 of the authenticated value (assumes TBI is off)
* compare the pointer value with the `XPAC`ed one
* assume CPU implements FEAT_FPAC and check nothing (or maybe ignore this for performance reason)

To me, it looks like we have to trust the contents of LR on function entry and hope that the caller provides a safe value to us, but there seems to be no reason to trust `x0`, for example.

jacobbramley replied:
There is _something_ special about `lr`, which is that it is written by `bl` or `blr`, and we usually expect that to be the return address at a function's entry point. With `pac-ret`, functions have to trust their provided `lr`:

```
fn:
    # If `lr` is bad on entry, it still signs and authenticates correctly.
    paciasp
    ...
    autiasp
    ret
```

Note that without `+leaf`, we don't sign or authenticate `lr` at all. The threat we're guarding against here is that the return address is overwritten between the function prologue and epilogue, not that the return address was invalid in the first place.

For the example highlighted by @atrosinenko, where `x1` is passed, then returned to, but never stored to memory, I think I'd prefer if it did raise an error.

- `f_nonx30_ret_non_auted` is a function that cannot be conventionally called with `bl` or `blr`, so is highly likely to be doing some weird control flow that justifies attention.
- It seems more likely that control-flow gadgets will be available to feed an attacker-controlled `x1` to a function than for `lr`.
- It's not AAPCS64, and arguably not `pac-ret`.

That said, it's correct that the scanner doesn't raise an error here:

```
fn:
    paciasp
    ...
    autia x0, sp
    ret x0
```

Although unusual, the fact that it returns to `x0` doesn't change anything in terms of trust. It still trusts `lr` on function entry.

To summarise, I think this scanner should accept `ret <reg>` where any of the following are true:
- `<reg>` is `lr` and has not been modified during the function.
- `<reg>` is authenticated.

This leaves false negatives like this:

```
fn:
    pacia x0, sp
    ...
    autia lr, sp
    ret
```

... but identifying that in general could be very difficult.

----

For forward-edge PAuth (like PAuth ABI), the threat model is slightly different, and attack could take place any time between pointer generation (e.g. by the runtime loader) and the point of use. For that reason, we'd have to authenticate every pointer, and that'd require a second `RegsToTrackInstsFor`, but it's currently out of scope of this scanner.

kbeyls replied:
Thank you both for the interesting discussion!
I do like the observation that the pac-ret hardening scheme implies that `x30` must be a non-signed value at function entry, and therefore, the pac-ret hardening scheme implicitly assumes that the caller is responsible for making sure `x30` is not attacker-controlled.

I wonder if for other hardening schemes, such as PAuthABI, a general rule could be found for which registers are expected to be "guaranteed not controlled by an attacker and passed raw in-register" versus "expected to be passed as a signed value".

I think figuring that out should be part of another pull request/discussion.

In the latest commit([7a12544](https://github.com/llvm/llvm-project/pull/122304/commits/7a125445d56da00db73165948dfc5f809644734d)), I did add 2 "FIXME" test cases with some extended comment trying to summarize some of the key points in the discussion above.

kbeyls committed:
Add FIXME test cases for passing potentially attacker-controlled register values across functions.

atrosinenko reviewed (commented):
Sorry for the delay, added several comments on `NonPacProtectedRetAnalysis.cpp` file, in other respects the PR looks good to me as to a reviewer of the second kind ("People familiar with the pac-ret hardening scheme and AArch64").

kbeyls committed:
More nitpicks based on feedback from atrosinenko

kbeyls commented on bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:
```diff
@@ -0,0 +1,952 @@
+// performs a tail call, that x30 might be attacker-controlled.
+// CHECK-NOT: function f_detect_clobbered_x30_passed_to_other
+        b   f_tail_called
+        .size f_callclobbered_x30, .-f_callclobbered_x30
```
Good catch! Now fixed.

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,522 @@
+raw_ostream &operator<<(raw_ostream &OS, const GeneralDiagnostic &Diag) {
+  OS << "diag<'" << Diag.Text << "'>";
+  return OS;
+}
```
it was unused indeed, now removed.

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,522 @@
+
+  SmallPtrSet<const MCInst *, 4> &lastWritingInsts(State &S,
+                                                   MCPhysReg Reg) const {
+    assert(Reg < Reg2StateIdx.size());
```
I think `isTrackingReg(Reg)` checks something different than `Reg < Reg2StateIdx.size()`. But you're right that we could also `assert(isTrackingReg(Reg))`.
I don't think `Reg2StateIdx[Reg]` is good to check for, as there will be one register that is tracked for which `RegStateIdx[Reg]` is 0?

In the latest commit I ended up adding `assert(isTrackingReg(Reg));`

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,522 @@
+  OS << ", at address " << llvm::format("%x", RetInst.getAddress()) << "\n";
+  OS << "  The return instruction is ";
+  BC.printInstruction(OS, RetInst, RetInst.getAddress(), BF);
+}
```
There is quite a lot of duplication indeed between `Gadget::generateReport` and `GenDiag::generateReport`.
Looking at it, I thought it was slightly more elegant to let `Gadget::generateReport` construct a `GenDiag` and call `generateReport` on it for the shared part of the diagnostic format, rather than factoring it out to base class `Annotation`.
That said, I'm not opposed to moving it to `Annotation` in the future, if/when we add more `diagnostics` classes and that would simplify the overall structure....

atrosinenko replied:
Looks good, thank you!

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,522 @@
+    if (BC.MIB->isCall(Point))
+      Written.set();
+    else
+      BC.MIB->getWrittenRegs(Point, Written);
```
I'm not fully sure here.
I wonder if you could think of a test case (maybe from pauthabi rather than pac-ret), where this could make a difference?
Should we maybe just add a FIXME here and only make changes when there is a test case that shows the difference in analysis between using `getWrittenRegs` and `getClobberedRegs`?

kbeyls replied:
Thanks, I added a FIXME for this now.

kbeyls commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,522 @@
+
+    ErrorOr<MCPhysReg> AutReg = BC.MIB->getAuthenticatedReg(Point);
+    if (AutReg && *AutReg != BC.MIB->getNoRegister()) {
+      Next.NonAutClobRegs.reset(
+          BC.MIB->getAliases(*AutReg, /*OnlySmaller=*/true));
+      if (TrackingLastInsts && isTrackingReg(*AutReg))
+        lastWritingInsts(Next, *AutReg).clear();
```
Similary to the above comment, I'm inclined to not change the code unless there is a test case that shows the difference. I wonder if you happen to have a test case or an idea for a test case?

atrosinenko replied:
After experimenting a bit, I have not managed to trick this code into reporting an incorrect set of last writing instructions. Anyway, even if an issue exists here, it should not be too severe as it only influences additional diagnostic information and not whether the report is generated or not.

kbeyls replied:
I also added a short FIXME here.

atrosinenko commented on bolt/lib/Passes/NonPacProtectedRetAnalysis.cpp:
```diff
@@ -0,0 +1,522 @@
+      };
+
+  ParallelUtilities::PredicateTy SkipFunc = [&](const BinaryFunction &BF) {
+    return false; // BF.shouldPreserveNops();
```
This seems to be resolved as well.

kbeyls committed:
Add 2 FIXME comments

kbeyls commented:
I think we're converging on having this PR ready to be merged.
@maksfb : before merging I thought to just check if you see any reason not to merge this PR?

maksfb commented:
> I think we're converging on having this PR ready to be merged. @maksfb : before merging I thought to just check if you see any reason not to merge this PR?

The overall structure of the PR looks good to me. Thanks!

kbeyls commented:
I merged this PR a few hours ago. Thank you for the extensive reviews @atrosinenko , @jacobbramley and @smithp35 !

alanzhao1 commented:
FYI I bisected a test failure to this change:

```
Testing:
 FAIL: BOLT :: binary-analysis/AArch64/gs-pacret-autiasp.s (1657 of 88694)
 ******************** TEST 'BOLT :: binary-analysis/AArch64/gs-pacret-autiasp.s' FAILED ********************
 Exit Code: 1
 
 Command Output (stderr):
 --
 RUN: at line 1: /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/bin/clang  --target=x86_64-unknown-linux-gnu -fPIE -fuse-ld=lld -Wl,--unresolved-symbols=ignore-all -Wl,--build-id=none -pie --target=aarch64-linux-gnu -nostartfiles -nostdlib -ffreestanding -march=armv9.5-a+pauth-lr -mbranch-protection=pac-ret /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/../../Inputs/asm_main.c -o /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/tools/bolt/test/binary-analysis/AArch64/Output/gs-pacret-autiasp.s.tmp.exe
 + /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/bin/clang --target=x86_64-unknown-linux-gnu -fPIE -fuse-ld=lld -Wl,--unresolved-symbols=ignore-all -Wl,--build-id=none -pie --target=aarch64-linux-gnu -nostartfiles -nostdlib -ffreestanding -march=armv9.5-a+pauth-lr -mbranch-protection=pac-ret /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/../../Inputs/asm_main.c -o /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/tools/bolt/test/binary-analysis/AArch64/Output/gs-pacret-autiasp.s.tmp.exe
 ld.lld: warning: cannot find entry symbol _start; not setting start address
 RUN: at line 2: /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/bin/llvm-bolt-binary-analysis --scanners=pacret /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/tools/bolt/test/binary-analysis/AArch64/Output/gs-pacret-autiasp.s.tmp.exe 2>&1 | /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/bin/FileCheck /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s
 + /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/bin/FileCheck /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s
 + /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/bin/llvm-bolt-binary-analysis --scanners=pacret /b/s/w/ir/cache/builder/src/third_party/llvm-build/Release+Asserts/tools/bolt/test/binary-analysis/AArch64/Output/gs-pacret-autiasp.s.tmp.exe
 /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s:16:17: error: CHECK-LABEL: expected string not found in input
 // CHECK-LABEL: GS-PACRET: non-protected ret found in function f1, basic block .LBB{{[0-9]+}}, at address
                 ^
 <stdin>:1:1: note: scanning from here
 BOLT-INFO: shared object or position-independent executable detected
 ^
 <stdin>:8:1: note: possible intended match here
 GS-PACRET: non-protected ret found in function f1, basic block .Ltmp0, at address 10270
 ^
 
 Input file: <stdin>
 Check file: /b/s/w/ir/cache/builder/src/third_party/llvm/bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s
 
 -dump-input=help explains the following input dump.
 
 Input was:
 <<<<<<
             1: BOLT-INFO: shared object or position-independent executable detected
 label:16'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
             2: BOLT-INFO: Target architecture: aarch64
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             3: BOLT-INFO: BOLT version: 72768d9bb8ad3e97a852270726f04d7167d9ef50
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             4: BOLT-INFO: first alloc address is 0x0
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             5: BOLT-INFO: creating new program header table at address 0x200000, offset 0x200000
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             6: BOLT-WARNING: non-relocation mode for AArch64 is not fully supported
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             7:  
 label:16'0     ~
             8: GS-PACRET: non-protected ret found in function f1, basic block .Ltmp0, at address 10270
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 label:16'1     ?                                                                                        possible intended match
             9:  The return instruction is 00010270: ret
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            10:  The 1 instructions that write to the return register after any authentication are:
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            11:  1. 0001026c: ldp x29, x30, [sp], #0x10
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            12:  This happens in the following basic block:
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            13:  00010268: add x0, x0, #0x3
 label:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             .
             .
             .
 >>>>>>
 
 --
 
 ********************
 
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..
 ********************
 Failed Tests (1):
   BOLT :: binary-analysis/AArch64/gs-pacret-autiasp.s
```

kbeyls commented:
> FYI I bisected a test failure to this change:

Thanks for reporting this.
From the error message, it looks like there is some non-determinism in how BOLT names basic blocks. In all of the test runs I've done on a few systems, the name of the basic block for this test always matched `.LBB{{[0-9]+}}`. In the failure you shared, it seems the basic block was named `.Ltmp0`. I'll be updating the test cases shortly to a regular expression that will match both styles of basic blocks..
